src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs

```diff
@@ -208,25 +208,57 @@ static uint GetTempPathW(int bufferLen, ref char buffer)
         {
-            var tempPathBuilder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
-
-            GetTempPath(ref tempPathBuilder);
-
-            var builder = new ValueStringBuilder(stackalloc char[PathInternal.MaxShortPath]);
+            // Avoid GetTempFileNameW because it is limited to 0xFFFF possibilities, which both
+            // means that it may have to make many attempts to create the file before
+            // finding an unused name, and also that if an app "leaks" such temp files,
+            // it can prevent GetTempFileNameW succeeding at all.
+            //
+            // To make this a little more robust, generate our own name with more
+            // entropy. We could use GetRandomFileName() here, but for consistency
+            // with Unix and to retain the ".tmp" extension we will use the "tmpXXXXXX.tmp" pattern.
+            // Using 32 characters for convenience, that gives us 32^^6 ~= 10^^9 possibilities,
+            // but we'll still loop to handle the unlikely case the file already exists.
+
+            const int KeyLength = 4;
+            byte* bytes = stackalloc byte[KeyLength];
+
+            Span<char> span = stackalloc char[13]; // tmpXXXXXX.tmp
+            span[0] = span[10] = 't';
+            span[1] = span[11] = 'm';
+            span[2] = span[12] = 'p';
+            span[9] = '.';
+
+            int i = 0;
+            while (true)
+            {
+                Interop.GetRandomBytes(bytes, KeyLength);  // 4 bytes = more than 6 x 5 bits
 
-            uint result = Interop.Kernel32.GetTempFileNameW(
-                ref tempPathBuilder.GetPinnableReference(), "tmp", 0, ref builder.GetPinnableReference());
+                byte b0 = bytes[0];
+                byte b1 = bytes[1];
+                byte b2 = bytes[2];
+                byte b3 = bytes[3];
 
-            tempPathBuilder.Dispose();
+                span[3] = (char)Base32Char[b0 & 0b0001_1111];
+                span[4] = (char)Base32Char[b1 & 0b0001_1111];
+                span[5] = (char)Base32Char[b2 & 0b0001_1111];
+                span[6] = (char)Base32Char[b3 & 0b0001_1111];
+                span[7] = (char)Base32Char[((b0 & 0b1110_0000) >> 5) | ((b1 & 0b1100_0000) >> 3)];
+                span[8] = (char)Base32Char[((b2 & 0b1110_0000) >> 5) | ((b3 & 0b1100_0000) >> 3)];
 
-            if (result == 0)
-                throw Win32Marshal.GetExceptionForLastWin32Error();
+                string path = string.Concat(Path.GetTempPath(), span);
 
-            builder.Length = builder.RawChars.IndexOf('\0');
+                try
+                {
+                    File.OpenHandle(path, FileMode.CreateNew, FileAccess.Write).Dispose();
+                }
+                catch (IOException ex) when (i < 100 && Win32Marshal.TryMakeWin32ErrorCodeFromHR(ex.HResult) == Interop.Errors.ERROR_FILE_EXISTS)
+                {
+                    i++; // If there's a read-only temp volume, don't loop forever
```

Comment:
If there's a read-only temp, the error will be ERROR_FILE_EXISTS?

Reply:
I was imagining a read only temp in which the file exists. Perhaps it would be reasonable to just throw in this case -- I can remove the counter? I'm just wary of hanging in some situation I can't think of.

Reply:
If someone's created ~1bn files through this API, it will hang (if I remove the counter) .. is that something we can imagine being plausible, some years from now? I guess we could fix it then?

I'm inclined to leave the counter in (I can remove the comment) just let me know what you prefer.

Reply:
I don't mind the counter.  I just don't understand the comment and why it's specific to a read-only temp.  The comment is suggesting that trying to create a file on a read-only temp will produce ERROR_FILE_EXISTS.  And if that's not the case and the file does exist, then how is the readonly-ness relevant?

Reply:
I think the comment can be deleted now that we are checking that the error is `ERROR_FILE_EXISTS`. We won't loop at all for read only temp anymore.